### DISCUSSION TITLE
Help needed to create and then overwrite a S3 BucketObject in the same session

### DISCUSSION THREAD

**Discussion by GeXiaoguo** [pulumi/pulumi#7309] [Pulumi Providers] [answered]
I am piloting with Pulumi and need some help with the following use case.

Suppose I create a S3 bucket and loads all the files from a local folder. When that is done, I proceed to overwrite one of the file uploaded with a deployment time value. Pulumi will give me a _"Duplicate resource URN"_ error. Is there any way this can be done.

I've added some demo code below.

As part of local file upload, the following bucket object is created.

```
           var runtime-configJS = new BucketObject("runtime-config.js", new BucketObjectArgs
           {
               Bucket = bucket.BucketName,
               Source = new FileAsset(filePath),
           });
```
After a deployment value is available, I try to overwrite the object content with code below.
```
           var runtimeConfigJSNew = new BucketObject("runtime-config.js", new BucketObjectArgs
           {
               Bucket = bucket.BucketName,
               Content = "config: deployment-value",
           });
```

**Comment by stack72** [comment]
Hi @GeXiaoguo 

I don't quite understand the flow here - are you manually changing that value between Pulumi runs or as part of the same configuration?

Paul

**Comment by GeXiaoguo** [comment, marked as answer]
I am changing the value as part of the same configuration.

To add some context, I am piloting a infrastructure as code solution with Pulumi. I've documented my steps here: https://github.com/GeXiaoguo/pulumi-s3-static-web/blob/master/readme-consuming-pulumi-output.md
To summarize, I am deploying the following resources
1. S3 backed front end(i.e. index.html + runtime-config.js)
where runtime-config.js is a file from a local folder and has the content below
```
window['runtime-config'] = {apiUrl: 'http://localhost:8080/api'}
```
2. AWS API Gateway + Lambda for backend
3. Overwrite the runtime-config.js file in the S3 bucket with the following code
```
var url = CreateRestAPIGatewayResources(apiGatewayLambda);
var runtimeConfigJS = url.Apply(x=> new BucketObject("runtime-config.js", new BucketObjectArgs
{
    Bucket = bucket.BucketName,
    Content = $@"window['runtime-config'] = {{apiUrl: '${x}'}}",
}));
```
All 3 steps are in the same Pulumi run where I've got the `Duplicated resource URN` error

**Reply by tenwit** [reply]
The issue is that your `new BucketObject()` is creating a new object, and you want to modify the existing one. Since Pulumi is declarative, it isn't possible _within a single call to `pulumi up`_ to define the same object twice, or follow an imperative flow where you define an object and update it.
Your options are

1. to exclude the object from your initial upload, and instead create the correctly-defined object after uploading all the other files, or
2. to use automation-api to separate the work into two calls to `pulumi up`, one which creates the bucket and uploads the local files, and another which changes the state of runtime-config.js.

**Comment by GeXiaoguo** [comment]
B.t.w If overwriting object is possible in the same Pulumi run, I believe it could also be used as a workaround for [this issue](https://github.com/pulumi/pulumi/issues/3021), which I am also facing at the moment. 

For a typical Serverless Web APP, the following dependencies are very common.

- the frontend (runtime-config.js in my case) needs to know the API Gateway url
- the API Gateway needs to know the front end domain to setup CORS

**Comment by katcosgrove** [comment]
GitHub discussions doesn't let you mark a subthread as the answer, so commenting to say @tenwit is correct with the answer [in this reply](https://github.com/pulumi/pulumi/discussions/7309#discussioncomment-875408)!
